File: test/data/models/check_response_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:libcheck/data/models/check_response.dart';

void main() {
  group('CheckResponse', () {
    test('parses complete response with continue=0', () {
      final json = {
        'session': 'abc123',
        'continue': 0,
        'books': {
          '9784774142230': {
            'Tokyo_Minato': {
              'status': 'OK',
              'reserveurl': 'https://example.com/reserve',
              'libkey': {
                'みなと': '貸出可',
                '三田': '貸出中',
              },
            },
          },
        },
      };

      final response = CheckResponse.fromJson(json);

      expect(response.session, 'abc123');
      expect(response.continueFlag, 0);
      expect(response.books, hasLength(1));

      final bookStatus = response.books['9784774142230']!['Tokyo_Minato']!;
      expect(bookStatus.status, 'OK');
      expect(bookStatus.reserveUrl, 'https://example.com/reserve');
      expect(bookStatus.libKeys, {'みなと': '貸出可', '三田': '貸出中'});
    });

  });

  group('BookSystemStatus', () {
    test('parses from JSON with reserveurl', () {
      final json = {
        'status': 'OK',
        'reserveurl': 'https://example.com/reserve',
        'libkey': {'みなと': '貸出可'},
      };

      final status = BookSystemStatus.fromJson(json);

      expect(status.status, 'OK');
      expect(status.reserveUrl, 'https://example.com/reserve');
      expect(status.libKeys, {'みなと': '貸出可'});
    });

    test('parses empty string reserveurl as null', () {
      final json = {
        'status': 'OK',
        'reserveurl': '',
        'libkey': {'みなと': '貸出可'},
      };

      final status = BookSystemStatus.fromJson(json);

      expect(status.reserveUrl, isNull);
    });
  });
}

File: src/data/models/checkResponse.test.ts
import { describe, expect, test } from 'vitest';

import {
  bookSystemStatusFromJson,
  checkResponseFromJson,
} from '@/data/models/checkResponse';

describe('CheckResponse', () => {
  test('parses complete response with continue=0', () => {
    const json = {
      session: 'abc123',
      continue: 0,
      books: {
        '9784774142230': {
          Tokyo_Minato: {
            status: 'OK',
            reserveurl: 'https://example.com/reserve',
            libkey: {
              みなと: '貸出可',
              三田: '貸出中',
            },
          },
        },
      },
    };

    const response = checkResponseFromJson(json);

    expect(response.session).toBe('abc123');
    expect(response.continueFlag).toBe(0);
    expect(Object.keys(response.books)).toHaveLength(1);

    const bookStatus = response.books['9784774142230']!.Tokyo_Minato!;
    expect(bookStatus.status).toBe('OK');
    expect(bookStatus.reserveUrl).toBe('https://example.com/reserve');
    expect(bookStatus.libKeys).toEqual({ みなと: '貸出可', 三田: '貸出中' });
  });
});

describe('BookSystemStatus', () => {
  test('parses from JSON with reserveurl', () => {
    const json = {
      status: 'OK',
      reserveurl: 'https://example.com/reserve',
      libkey: { みなと: '貸出可' },
    };

    const status = bookSystemStatusFromJson(json);

    expect(status.status).toBe('OK');
    expect(status.reserveUrl).toBe('https://example.com/reserve');
    expect(status.libKeys).toEqual({ みなと: '貸出可' });
  });

  test('parses empty string reserveurl as null', () => {
    const json = {
      status: 'OK',
      reserveurl: '',
      libkey: { みなと: '貸出可' },
    };

    const status = bookSystemStatusFromJson(json);

    expect(status.reserveUrl).toBeUndefined();
  });
});
